scipp_widgets/widgets.py

```diff
@@ -0,0 +1,146 @@
+        """
+        super().__init__()
+        self.layout.flex_flow = 'column'
+        self.callable = callable
+        self.input_specs = input_specs
+        self.input_widgets = []
+        self._setup_input_widgets(input_specs)
+
+        self.button = widgets.Button(description=name)
+        self.button.on_click(self._on_button_clicked)
+        self.button_widgets = [self.button]
+        if (hide_code):
+            self.button_widgets += (HideCodeWidget(True), )
+
+        self.output_area = widgets.Output()
+        self.output_widgets = widgets.VBox([self.output_area])
+
+        self.row_widgets = widgets.HBox(self.input_widgets +
+                                        self.button_widgets)
+        self.row_widgets.layout.flex_flow = 'row wrap'
+
+        self.children = [self.row_widgets, self.output_widgets]
+
+    def _setup_input_widgets(self, inputs):
+        """
+        Creates a user-input widget for each item in inputs
+        """
+        for spec in inputs:
+            self.input_widgets.append(spec.create_input_widget())
+
+    def _retrive_kwargs(self):
+        kwargs = {
+            spec.name: spec.validate(widget.value)
+            for spec, widget in zip(self.input_specs, self.input_widgets)
+        }
+        return kwargs
+
+    def _on_button_clicked(self, button):
+        self.output_area.clear_output()
+        with self.output_area:
+            try:
+                kwargs = self._retrive_kwargs()
+            except ValueError as e:
+                print(f'Invalid inputs: {e}')
+                return
+
+            self._process(kwargs)
+
+    def _process(self, kwargs):
+        display(self.callable(**kwargs))
+
+
+class ProcessWidget(DisplayWidget):
+    """
+    Provides a simple graphical wrapper around a given callable.
+    """
+    def __init__(self, callable, name: str, inputs, hide_code=False):
+        """
+        Parameters:
+        callable (Callable): The function to call
+        name (str): name of widget to display
+        inputs (list): class containing input data
+        """
+        super().__init__(callable, name, inputs, hide_code=hide_code)
+        self.scope = get_notebook_global_scope()
+
+        self.output = widgets.Text(placeholder='output name',
+                                   value='',
+                                   continuous_update=False)
+        self.row_widgets.children = self.input_widgets + [
+            self.output
+        ] + self.button_widgets
+
+    def _process(self, kwargs):
+        """
+        Calls the wrapped function using the
+        parameter values specified.
+
+        Returns: Output of the wrapped callable.
+        """
+        if self.output.value:
+            output_name = self.output.value
+        else:
+            print('Invalid inputs: No output name specified')
+            return
+
+        self.scope[output_name] = self.callable(**kwargs)
+        display(self.scope[output_name])
```

Comment:
Why call `display` in a processing widget?

Reply:
This is so the output of the operation is displayed below the widgets when the button is clicked. I feel this makes it clearer both when the operation has finished and what it has produced but is definitely optional if this isn't deemed useful.

Reply:
I am fine with leaving it in for now, so we can try it out.